core/src/fields.rs

```diff
@@ -569,34 +575,45 @@ impl MongoFields {
 impl MongoStatement for MongoFields {
     // Move the cursor to the next document and update the current row.
     // Return true if moving was successful, false otherwise.
-    fn next(&mut self, mongo_connection: Option<&MongoConnection>) -> Result<bool> {
+    fn next(
+        &mut self,
+        mongo_connection: Option<&MongoConnection>,
+    ) -> Result<(bool, Option<Error>)> {
         match self.field_name_filter.as_ref() {
             None => {
                 self.current_field_for_collection += 1;
-                Ok(
-                    (self.current_field_for_collection as usize) < self.current_col_metadata.len()
-                        || self.get_next_metadata(mongo_connection.unwrap()),
-                )
+                match (self.current_field_for_collection as usize) < self.current_col_metadata.len()
+                {
+                    true => Ok((true, None)),
+                    false => self.get_next_metadata(mongo_connection.unwrap()),
+                }
             }
             Some(filter) => {
                 let filter = filter.clone();
                 loop {
                     self.current_field_for_collection += 1;
-                    if (self.current_field_for_collection as usize
-                        >= self.current_col_metadata.len())
-                        && !self.get_next_metadata(mongo_connection.unwrap())
+                    let mut e: Option<Error> = None;
+                    if self.current_field_for_collection as usize >= self.current_col_metadata.len()
                     {
-                        return Ok(false);
+                        let next = self.get_next_metadata(mongo_connection.unwrap());
+                        if let Ok((b, error)) = next {
+                            // store the first error we see from get_next_metadata, to return at conclusion of function
+                            if e.is_none() && error.is_some() {
+                                e = error;
```

Comment:
If you change get_next_metadata to return a vector, this will have to change to a vector too.
And you will also have to add a vector of errors to the existing one if this is not the first error.
Also, I think there is a missing "continue" because if `get_next_metadata` comes back with an error, then trying to do filter will likely likely fail because `self.current_field_for_collection as usize >= self.current_col_metadata.len()`, no?

Reply:
Thanks for the heads about changes if errors becomes a vector. I could be mistaken, but I think even if we get errors in `get_next_metadata`, it will still continue until it successfully parses a collection and resets the `current_field_for_collection` (or runs out of available collections, and returns `Ok(false, ...)`, which might eliminate the need for a continue

Reply:
ohhh, you are right, for some reason I thought that the filter was inside the loop. Sorry about that.